lib/spack/spack/fetch_strategy.py

```diff
@@ -1007,9 +1009,18 @@ def clone(self, dest=None, commit=None, branch=None, tag=None, bare=False):
                     git(*args)
 
         # Init submodules if the user asked for them.
-        if self.submodules:
+        submodules = self.submodules
+        if self.package is not None and hasattr(self.package, "submodules"):
+            submodules = self.package.submodules() or submodules
+        if submodules:
             with working_dir(dest):
-                args = ['submodule', 'update', '--init', '--recursive']
+                args = ['submodule', 'init']
+                if isinstance(submodules, (list, tuple)):
```

Comment:
This looks like a bug waiting to happen: what the user accidentally returns a string? Then it gets ignored instead of generating an error. Maybe just 
```suggestion
                if submodules:
```
and raise an assertion if it's not a list/tuple/NoneType?

Reply:
If this were being designed from the ground up, maybe this would be better, but it will break backward compatibility.  Users currently tell Spack to fetch a package's submodules with the `submodules=` flag to its `version` declarations.  Spack merely checks the truthiness of this value - ie, spack currently treats `version("blah", submodules=True)`, `version("blah", submodules="baz")`, `version("blah", submodules="false")`, etc as equivalent.  I guess one way around it would be to define in the `PackageBase` something like:

```python
def get_submodules(self):
    submodules = self.submodules()
    if submodules is None:
        return
    elif isinstance(submodules, six.string_types):
        return shlex.split(submodules)
    elif instance(submodules, (list, tuple):
        return list(submodules)
    else:
        raise ValueError("Expected Package.submodules() to return a list")
```

and have the fetcher call `PackageBase.get_submodules()`